src/wallet/wallet.cpp

```diff
@@ -1288,24 +1288,11 @@ void CWallet::UpdatedBlockTip()
 
 void CWallet::BlockUntilSyncedToCurrentChain() {
     AssertLockNotHeld(cs_wallet);
-
-    {
-        // Skip the queue-draining stuff if we know we're caught up with
-        // chainActive.Tip()...
-        // We could also take cs_wallet here, and call m_last_block_processed
-        // protected by cs_wallet instead of cs_main, but as long as we need
-        // cs_main here anyway, it's easier to just call it cs_main-protected.
-        auto locked_chain = chain().lock();
-
-        if (!m_last_block_processed.IsNull() && locked_chain->isPotentialTip(m_last_block_processed)) {
-            return;
-        }
-    }
-
-    // ...otherwise put a callback in the validation interface queue and wait
+    // Skip the queue-draining stuff if we know we're caught up with
+    // chainActive.Tip(), otherwise put a callback in the validation interface queue and wait
     // for the queue to drain enough to execute it (indicating we are caught up
     // at least with the time we entered this function).
-    chain().waitForNotifications();
+    chain().waitForNotificationsUpToTip(m_last_block_processed);
```

Comment:
Accessing `m_last_block_processed` requires the `cs_wallet` lock. I think to make this work you'll need to:
- grab `cs_wallet`
- access `m_last_block_processed`
- release `cs_wallet`
- call the `waitForNotificationsUpToTip` interface method

Alternatively, we could make `m_last_block_processed` an atomic.

Thoughts, @ryanofsky ?

Reply:
Yes, John's right this needs to synchronize access to m_last_block_processed. Since I don't think there's a straightforward way to make uint256 atomic, maybe a clean way to do this would be to add a new `sync.h` macro:

```diff
-chain().waitForNotificationsUpToTip(m_last_block_processed);
+chain().waitForNotificationsUpToTip(WITH_LOCK(cs_wallet, return m_last_block_processed));
```

```c++
//! Run code while locking a mutex.
//!
//! Examples:
//!
//!   WITH_LOCK(cs, shared_val = shared_val + 1);
//!
//!   int val = WITH_LOCK(cs, return shared_val);
//!
#define WITH_LOCK(cs, code) [&]{ LOCK(cs); code; }()
```

Reply:
I like the macro, but I don't like putting the complete expression in the `waitForNotificationsUpToTip()` function call. It's too easy for a casual observer to think that the lock is held for the entire call to `waitForNotificationsUpToTip()` (which of course would cause a deadlock). How about:

```c++
uint256 last_block_hash = WITH_LOCK(cs_wallet, m_last_block_processed);
chain().waitForNotificationsUpToTip(last_block_hash);
```

which perhaps compiles to the same thing.

Reply:
Ah yes, wasn't sure of cs_wallet lock covering. Added a WITH_LOCK commit and used to get back m_last_block_processed as showed in example.